### ISSUE TITLE
Updating deepchem.feat.graph_data.GraphData

### ISSUE THREAD

**Issue by arunppsg** [deepchem/deepchem#2824]
## 🚀 Feature

DeepChem currently uses `dc.feat.graph_data.GraphData` for handling graphs. I think it will be more apt for `GraphData` to be under the `dc.data`, thereby making them a fundamental data type supported by DeepChem.

### Motivation
- Graph based data types seems to be a very common kind of data type used in scientific applications.
- We can add more support for handling graphs and performing operations on them - like saving, loading, converting it to other formats etc

### Additional Context on transition

For doing the transition, we can probably begin with createing `dc.data.GraphDataset` class like `dc.data.NumpyDataset` which can take in node features and edge features - just like `dc.feat.graph_data.GraphData` but under `dc.data`. Next, we can add a depreciation note and update code in repository to use dc.data.GraphDataset. Later on, we can add more support on handling (like saving, retrieving) graphs.

Some features which will be nice to see (feel free to add yours):
- Conversion of pytorch-geometric, networkx and other graph forms to dc.data.GraphDataset and vice-versa
- Utilities to compute graph properties like centrality measures, etc

**Comment by arunppsg** [comment]
Also, most graph models in DeepChem takes `DGLGraph` as input. Having `dc.data.GraphDataset` can allow for more customization as well as compatibility with other formats like `torch_geometric.data.Data` and `DGLGraph`.

**Comment by neerajanand-coder** [comment]
@arunppsg I would like to work on this issue.

**Comment by arunppsg** [comment]
Yes, that will be great @neerajanand-coder . 

After discussion from weekly dev-call, we can start by moving `dc.feat.GraphData` to `dc.data.GraphData`. 

Regarding creation of `dc.data.GraphDataset`, we need to think of what special utility/purposes it serves. We can plan it as a later project. In many scenarios, `dc.data.NumpyDataset` can hold graph objects as numpy arrays. So `GraphDataset` requirement should be thought of.

cc: @rbharath

**Comment by neerajanand-coder** [comment]
Thanks, @arunppsg.

**Comment by peastman** [comment]
I don't understand what the GraphDataset class would do.  As you say, you can already store any kind of object in a dataset.  Dataset subclasses differ in where they store the data (in memory for NumpyDataset, in .npy files for DiskDataset, in image files for ImageDataset), but they all provide a common interface that presents the data as numpy arrays.

For the same reason, GraphData seems like it belongs in `feat`, not in `data`.  It is, after all, a featurization.  The `dc.data` module is concerned with how data is stored and accessed, while being generally agnostic about what the data represents or how it will be used.  Data representations are what `dc.feat` is all about.

**Comment by arunppsg** [comment]
Now it makes sense on why we are having it in `feat`. Thanks for the explanation!

I have two more feature requests regarding `GraphData`:

- When we print a `GraphData` object, we get an output like `<deepchem.feat.graph_data.GraphData at 0x7fdf678865d0>`. Example:
```
import numpy as np

node_features = np.random.rand(5, 10)
edge_index = np.array([[0, 1, 2, 3, 4], [1, 2, 3, 4, 0]], dtype=np.int64)
graph = dc.feat.graph_data.GraphData(node_features=node_features, edge_index=edge_index)

print (graph)
```
Can we add a small utility to so that it prints the graph more useful, something like `GraphData(node_features=[5, 10], edge_index=[2, 5])`?

- The second one is to for storing global graph level features. A sample use case: for crystals we have crystal temperatures - the features which belong to the entire graph and not specific to any node nor edge. Can we add a variable to hold graph level features?

**Comment by peastman** [comment]
> Can we add a small utility to so that it prints the graph more useful

Good idea!

**Comment by arunppsg** [comment]
> Can we add a small utility to so that it prints the graph more useful

@neerajanand-coder , will you be interested in working on this?

**Comment by neerajanand-coder** [comment]
@arunppsg yes for sure.

**Comment by neerajanand-coder** [comment]
@arunppsg I have implemented the first feature and working on the second.

**Comment by arunppsg** [comment]
@peastman I have a question on using `GraphData` with pytorch models. Since GraphData stores node_features and others as numpy arrays, to pass it to a pytorch model, should we convert the attributes to tensor every time during call of forward method or add support to store tensors in GraphData? 

Another way we can do is convert it to pytorch geometric graph using `to_pyg_graph` method available in `GraphData` and use the pyg graph when passing it to torch models.

**Comment by peastman** [comment]
I think that relates to a much more general question.  All datasets store their data as numpy arrays, and they need to be converted to tensors for each batch.  Do we want to create a mechanism where data can get preprocessed into a form that's more efficient for a particular model?  It's an interesting idea.  The first step would be to profile and figure out whether the conversions are currently a significant cost.  How much benefit could we hope to get?

That assumes you'll be passing the same batch or sample to the model many times.  If you only do it once, the cost is the same either way.

**Comment by arunppsg** [comment]
> Do we want to create a mechanism where data can get preprocessed into a form that's more efficient for a particular model? 

Yes, because for each epoch we are currently converting numpy array to tensor. Maybe in long run, having a mechanism can be more efficient.
> The first step would be to profile and figure out whether the conversions are currently a significant cost. How much benefit could we hope to get?

I will try to profile and see about the costs. Once we have a clear picture, we can decide to take it forward or not.

**Comment by peastman** [comment]
> Yes, because for each epoch we are currently converting numpy array to tensor.

Assuming the dataset is too large to fit in memory at once, creating new tensors on each epoch is unavoidable.  The question is whether there's a more efficient way of getting from data stored on disk to something ready to pass to pytorch.
